### PR TITLE
Add strategy for Mali GPU in vectorization path and enable hoisting

### PR DESCRIPTION
**Pull request by ThomasRaoux**
Add launch strategy for Mali GPU and a set of new transformation needed
and specify a good tile and workgroup size to use based on experiment.
Also enable hoisting of transfer_ops for the vector path.

### CODE REVIEW

File: iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp [hunk 1]
```diff
@@ -385,6 +387,61 @@ static void populateTilingToSubgroupPatterns(
...
+      getThreadProcInfoFn,
+      {linalg::DistributionMethod::CyclicNumProcsEqNumIters,
+       linalg::DistributionMethod::CyclicNumProcsEqNumIters}};
+  patterns.insert<linalg::LinalgTilingPattern<linalg::MatmulOp>>(
+      context,
+      linalg::LinalgTilingOptions()
+          .setLoopType(linalg::LinalgTilingLoopType::ParallelLoops)
+          .setTileSizeComputationFunction(getInnerTileSizeFn)
+          .setDistributionOptions(subgroupDistributionOptions),
+      linalg::LinalgMarker(
```
Comment (antiagainst) [inline iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp:438]: Can use `getLinalgMatchAndReplaceMarker` to simplify this a bit now.

File: iree/compiler/Conversion/LinalgToSPIRV/test/pipeline_test.mlir [hunk 2]
```diff
@@ -1,41 +1,32 @@
...
-        {a_type = f16, b_type = f16, c_type = f32, k_size = 16 : i32,
-         m_size = 16 : i32, n_size = 16 : i32, result_type = f32,
-         scope = 3 : i32}],
-       max_compute_shared_memory_size = 49152 : i32,
-       max_compute_workgroup_invocations = 1024 : i32,
-       max_compute_workgroup_size = dense<[2147483647, 65535, 65535]> : vector<3xi32>,
-       subgroup_size = 32 : i32}>} {
+       SPV_KHR_storage_buffer_storage_class, SPV_KHR_variable_pointers]>,
+      ARM:IntegratedGPU,
+      {cooperative_matrix_properties_nv = [],
```
Comment (antiagainst) [inline iree/compiler/Conversion/LinalgToSPIRV/test/pipeline_test.mlir:15]: Can remove `cooperative_matrix_properties_nv` here. It's allowed to be missing.

**Review by antiagainst — CHANGES_REQUESTED** [review]
Awesome! Just a few nits.

File: iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp [hunk 3]
```diff
@@ -547,12 +624,29 @@ void LinalgTileAndFusePass::runOnOperation() {
...
+                                     std::move(canonicalizationPatterns));
         LLVM_DEBUG({
           llvm::dbgs() << "--- After Vector Unroll ---\n";
           funcOp.print(llvm::dbgs(), OpPrintingFlags().useLocalScope());
           llvm::dbgs() << "\n\n";
         });
       }
+
+      {
+        linalg::hoistViewAllocOps(funcOp);
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp:641]: Is this to move the alloc for workgroup memory to the start of the function?

I am planning to change how this works, and adapt the allocation function callback to just inesrt the allocation at the start of the function. Also making the deallocation function callback a no-op. That will match semantics of workgroup memory better. With that this shouldnt be needed right?

File: iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp [hunk 4]
```diff
@@ -547,12 +624,29 @@ void LinalgTileAndFusePass::runOnOperation() {
         OwningRewritePatternList vectorUnrollPatterns;
         populateVectorUnrollPatterns(context, vectorUnrollPatterns);
         applyPatternsAndFoldGreedily(funcOp, std::move(vectorUnrollPatterns));
+
+        OwningRewritePatternList canonicalizationPatterns;
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp:628]: Maybe we should move all the vectorization related patterns and application to its own function to make the main function a bit more readable?

File: iree/compiler/Conversion/LinalgToSPIRV/test/pipeline_test_cooperative_mat.mlir [hunk 5]
```diff
@@ -0,0 +1,57 @@
+// RUN: iree-opt -split-input-file -pass-pipeline="iree-codegen-linalg-to-spirv-pipeline{use-vectorization}" %s | IreeFileCheck %s
+// RUN: iree-opt -split-input-file -pass-pipeline="iree-codegen-linalg-to-spirv-pipeline{use-vectorization use-workgroup-memory}" %s | IreeFileCheck %s
+
+module attributes {
```
Comment (MaheshRavishankar) [inline iree/compiler/Conversion/LinalgToSPIRV/test/pipeline_test_cooperative_mat.mlir:4]: Can we merge this with pipeline_test.mlir ? We can use `-split-input-file` and `-prefix-check` to consolidate all pipeline tests?

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Nice! Few Minor comments.

**Comment by ThomasRaoux** [comment]
Please take another look.

**Review by antiagainst — APPROVED** [review]

File: iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp (hunk 3 above)
Comment (MaheshRavishankar) [reply iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp]: Sure. Please add a TODO here and assign to me?

File: iree/compiler/Conversion/LinalgToSPIRV/test/pipeline_test_cooperative_mat.mlir (hunk 5 above)
Comment (MaheshRavishankar) [reply iree/compiler/Conversion/LinalgToSPIRV/test/pipeline_test_cooperative_mat.mlir:4]: Ah I see you moved tests from pipeline_test.mlir to this one. Sorry didnt realize that. This makes sense now.

**Review by MaheshRavishankar — APPROVED** [review]

File: iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp (hunk 3 above)
Comment (ThomasRaoux) [reply iree/compiler/Conversion/LinalgToSPIRV/LinalgTileAndFusePass.cpp]: Added a TODO for this.

**Pull request merged**